server/src/main/java/org/elasticsearch/repositories/FinalizeSnapshotContext.java

```diff
@@ -106,8 +108,21 @@ public ClusterState updatedClusterState(ClusterState state) {
         final ClusterState updatedState = SnapshotsService.stateWithoutSnapshot(state, snapshotInfo.snapshot(), updatedShardGenerations);
         // Now that the updated cluster state may have changed in-progress shard snapshots' shard generations to the latest shard
         // generation, let's mark any now unreferenced shard generations as obsolete and ready to be deleted.
+
+        final Collection<IndexId> deletedIndices = updatedShardGenerations.deletedIndices.indices();
         obsoleteGenerations.set(
-            SnapshotsInProgress.get(updatedState).obsoleteGenerations(snapshotInfo.repository(), SnapshotsInProgress.get(state))
+            SnapshotsInProgress.get(updatedState)
+                .obsoleteGenerations(snapshotInfo.repository(), SnapshotsInProgress.get(state))
+                .entrySet()
+                .stream()
+                // We want to keep both old and new generations for deleted indices, so we filter them out here to avoid deletion.
+                // We need the old generations because they are what get recorded in the RepositoryData.
```

Comment:
Hmm this seems suspicious. We should be updating the shard generations for _existing_ shards in `RepositoryData` and discarding the previous values, even if the index isn't included in the snapshot. AIUI the tripping assertion you mentioned [in an earlier comment](https://github.com/elastic/elasticsearch/pull/128650#issuecomment-2921860183) related to generations for shards that were totally absent from `RepositoryData`. We should drop those shards entirely from `RepositoryData`, but update any ones that do exist.